src/Discord.hs

```diff
@@ -116,18 +117,29 @@ runDiscord opts = do
 -- | Runs the main loop 
 runDiscordLoop :: DiscordHandle -> RunDiscordOpts -> IO T.Text
 runDiscordLoop handle opts = do
-  resp <- liftIO $ writeRestCall (discordHandleRestChan handle) GetCurrentUser
+  resp <- startupRestCalls
   case resp of
     Left (RestCallInternalErrorCode c e1 e2) -> libError $
              "HTTP Error Code " <> T.pack (show c) <> " " <> TE.decodeUtf8 e1
                                                    <> " " <> TE.decodeUtf8 e2
     Left (RestCallInternalHttpException e) -> libError ("HTTP Exception -  " <> T.pack (show e))
     Left (RestCallInternalNoParse _ _) -> libError "Couldn't parse GetCurrentUser"
-    _ -> do me <- liftIO . runReaderT (try $ discordOnStart opts) $ handle
-            case me of
-              Left (e :: SomeException) -> libError ("discordOnStart handler stopped on an exception:\n\n" <> T.pack (show e))
-              Right _ -> loop
+    Right (user, app) -> do putMVar (cacheHandleCache (discordHandleCache handle)) (Right (createCache user app))
+                            me <- liftIO . runReaderT (try $ discordOnStart opts) $ handle
+                            case me of
+                              Left (e :: SomeException) -> libError ("discordOnStart handler stopped on an exception:\n\n" <> T.pack (show e))
+                              Right _ -> loop
  where
+   startupRestCalls :: IO (Either RestCallInternalException (User, FullApplication))
+   startupRestCalls = do resp <- writeRestCall (discordHandleRestChan handle) R.GetCurrentUser
+                         case resp of
+                           Left e -> pure (Left e)
+                           Right user -> do
+                             resp2 <- writeRestCall (discordHandleRestChan handle) R.GetCurrentApplication
+                             case resp2 of
+                               Left e -> pure (Left e)
+                               Right app -> pure (Right (user, app))
```

Comment:
Open to suggestions for a more haskell way to write this

Reply:
```haskell
do
  eUser <- writeRestCall (discordHandleRestChan handle) R.GetCurrentUser
  eApp <- writeRestCall (discordHandleRestChan handle) R.GetCurrentApplication
  pure $ (,) <$> eUser <*> eApp
```
would be an alternative way of writing this I think.

Another alternative would be to match on `(Right user, Right app)` and then have two further cases for when either is Left, which would be equivalent to what you have but with less nesting.